### ISSUE TITLE
rules_haskell should store all shared library dependencies in one place like CC rules

### ISSUE THREAD

**Issue by aherrmann** [tweag/rules_haskell#593]
When building dynamically linked Haskell binaries rules_haskell currently behaves differently than the builtin CC rules.

The builtin CC rules will store all dynamic libraries in a special global directory, such as `_solibs_k8`, and the output of `cc_binary` will contain an `RPATH` entry that links to the `_solibs_k8` directory under its `runfiles` path.

rules_haskell, on the other hand, will store dynamic Haskell libraries in paths according to their package name and add separate RPATH entries per Haskell dependency. C library dependencies are stored in a path according to the Haskell binary's package name and a require one RPATH entry in total.

For example, on Linux a `cc_binary` with two direct and two indirect shared library dependencies has a dynamic section of the following form (ignoring system library dependencies):
```
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Sdata_Slibourclibrary.so]
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Ssolibs-dir_Slib-c_Sliblib-c.so]
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Ssolibs-dir_Slib-b_Sliblib-b.so]
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Ssolibs-dir_Slib-a_Sliblib-a.so]
 0x000000000000001d (RUNPATH)            Library runpath: [
                                           $ORIGIN/../../_solib_k8/
                                         ]
```
I.e. one reference per transitive shared library dependency, and one global RPATH entry.

However, a `haskell_binary` with two direct and two indirect C library dependencies and one direct and two indirect Haskell library dependencies has a dynamic section of the following form (ignoring system library dependencies and prebuilt dependencies):
```
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Sdata_Sourclibrary_libourclibrary.so]
 0x0000000000000001 (NEEDED)             Shared library: [libtests_Ssolibs-dir_Slib-c_Slib-c_liblib-c.so]
 0x0000000000000001 (NEEDED)             Shared library: [libHStestsZSsolibs-dirZSHsCZSHsC-ghc8.4.4.so]
 0x0000000000000001 (NEEDED)             Shared library: [libHStestsZSsolibs-dirZSHsBZSHsB-ghc8.4.4.so]
 0x0000000000000001 (NEEDED)             Shared library: [libHStestsZSsolibs-dirZSHsAZSHsA-ghc8.4.4.so]
 0x000000000000001d (RUNPATH)            Library runpath: [
                                           $ORIGIN/../../../../../bazel-out/k8-fastbuild/bin/tests/solibs-dir:
                                           $ORIGIN/../../../../../bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsA:
                                           $ORIGIN/../../../../../bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsB:
                                           $ORIGIN/../../../../../bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsC:
                                           bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsC:
                                           bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsB:
                                           bazel-out/k8-fastbuild/bin/tests/solibs-dir/HsA:
                                         ]
```
I.e. one reference for each direct C library dependency, one reference for each transitive Haskell library dependency, one RPATH entry for C library dependencies, and two RPATH entries per transitive Haskell library dependency (one should be enough and the duplicate is likely a bug).

With complex Haskell targets the RPATH entries can quickly grow out of hand and cause slow start up times, or on MacOS exceed the MACH-O header size limitation.

rules_haskell should behave as the builtin CC rules and place all shared library dependencies into a single directory, requiring only a single RPATH entry.

Due to [limitations in the Starlark API](https://github.com/bazelbuild/bazel/issues/4581) it is not possible to store the shared libraries in the global `_solibs_k8` directory. However, a directory per binary target would already allow to have only one RPATH entry.

**Comment by aherrmann** [comment]
Looking at the runfiles directory of a Haskell binary, Bazel already places a `_solibs_k8` directory populated with links to the direct C library dependencies. It is unclear to me whether this is actively caused by rules_haskell, or something that Bazel does automatically.

**Comment by mjrussell** [comment]
👍 for this, I was trying hazel/rules_haskell in a fairly large codebase and tripped macho very quickly

**Comment by aherrmann** [comment]
@mjrussell Thanks for pointing to `ld-wrapper-macos`. We tried to inject it a while ago a fairly large Bazel project but the effort failed. At the time we tried to inject it on the Nix side of our build setup without making rules_haskell aware of its existence. The issue was then that `ld-wrapper-macos` produced additional library files, that Bazel was not aware of and so they didn't show up in the corresponding runfiles. I'm hoping that the `_solib` directory approach will be sufficient to eliminate the header size issue. If not we may need to revisit `ld-wrapper-macos`.
